be/src/util/frame_of_reference_coding.cpp

```diff
@@ -0,0 +1,377 @@
+}
+
+template<typename T>
+void ForEncoder<T>::put_batch(const T* in_data, size_t count) {
+    if (_buffered_values_num + count < ForCoding::FRAME_VALUE_NUM) {
+        copy_value(in_data, count);
+        _count += count;
+        return;
+    }
+
+    // 1. padding one frame
+    size_t padding_num = ForCoding::FRAME_VALUE_NUM - _buffered_values_num;
+    in_data = copy_value(in_data, padding_num);
+    bit_packing_one_frame_value(_buffered_values);
+
+    // 2. process frame by frame
+    size_t frame_size = (count - padding_num) / ForCoding::FRAME_VALUE_NUM;
+    for (size_t i = 0; i < frame_size; i++) {
+        // directly encode value to the bit_writer, don't buffer the value
+        _buffered_values_num = ForCoding::FRAME_VALUE_NUM;
+        bit_packing_one_frame_value(in_data);
+        in_data += ForCoding::FRAME_VALUE_NUM;
+    }
+
+    // 3. process remaining value
+    size_t remaining_num = (count - padding_num) % ForCoding::FRAME_VALUE_NUM;
+    if (remaining_num > 0) {
+        copy_value(in_data, remaining_num);
+    }
+
+    _count += count;
+}
+
+// todo(kks): improve this method by SIMD instructions
+
+// Use as few bit as possible to store a piece of integer data.
+// param[in] input: the integer list need to pack
+// param[in] in_num: the number integer need to pack
+// param[in] bit_width: how many bit we use to store each integer data
+// param[out] out: the packed result
+
+// For example:
+// The input is int32 list: 1, 2, 4, 8 and bit_width is 3
+// The output will be: 0001 0010 0100 1000
+template<typename T>
+void ForEncoder<T>::bit_pack(T *input, uint8_t in_num, int bit_width, uint8_t *output) {
+    T in_mask = 0;
+    int bit_index = 0;
+    *output = 0;
+    for (int i = 0; i < in_num; i++) {
+        in_mask = 1 << (bit_width - 1);
+        for (int k = 0; k < bit_width; k++) {
+            if (bit_index > 7) {
+                bit_index = 0;
+                output++;
+                *output = 0;
+            }
+            *output |= (((input[i] & in_mask) >> (bit_width - k - 1)) << (7 - bit_index));
+            in_mask >>= 1;
+            bit_index++;
+        }
+    }
+}
+
+template<typename T>
+void ForEncoder<T>::bit_packing_one_frame_value(const T* input) {
+    T min = input[0];
+    T max = input[0];
+    bool is_ascending = true;
+    uint8_t bit_width = 0;
+
+    for (uint8_t i = 1; i < _buffered_values_num; ++i) {
+        if (is_ascending) {
+            if (input[i] < input[i - 1]) {
+                is_ascending = false;
+            } else {
+                bit_width = std::max(bit_width, bits(input[i]- input[i - 1])) ;
+            }
+        }
+
+        if (input[i] < min) {
+            min = input[i];
+            continue;
+        }
+
+        if (input[i] > max) {
+            max = input[i];
+        }
+    }
+
+    if (sizeof(T) == 8) {
+        put_fixed64_le(_buffer, min);
+    } else {
+        put_fixed32_le(_buffer, min);
+    }
+
+    T delta_values[_buffered_values_num];
```

Comment:
`_buffered_values_num` is only known at runtime. Here you use a compile extension called [VLA](https://en.wikipedia.org/wiki/Variable-length_array) which is not part of C++ standard. I think we should avoid such usage. In this case, you can reuse `_buffered_values` to store deltas.